src/runtime/xrt.cpp

```diff
@@ -0,0 +1,628 @@
+
+    if (create && (global_device == nullptr)) {
+        int status = create_xrt_context(user_context);
+        if (status != halide_error_code_success) {
+            halide_mutex_unlock(&thread_lock);
+            return status;
+        }
+    }
+
+    *instance_ret = global_instance;
+    *adapter_ret = global_adapter;
+    *device_ret = global_device;
+
+    return halide_error_code_success;
+}
+
+WEAK int halide_xrt_release_context(void *user_context) {
+    halide_mutex_unlock(&thread_lock);
+    return halide_error_code_success;
+}
+
+}  // extern "C" linkage
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+namespace XRT {
+
+// Helper object to acquire and release the XRT context.
+class XRTContext {
+    void *user_context;
+
+public:
+    void *instance = nullptr;
+    void *adapter = nullptr;
+    xrtDeviceHandle device = nullptr;
+    void *queue = nullptr;
+
+    int error_code = 0;
+
+    ALWAYS_INLINE XRTContext(void *user_context)
+        : user_context(user_context) {
+        error_code = halide_xrt_acquire_context(
+            user_context, &instance, &adapter, &device);
+        if (error_code == halide_error_code_success) {
+            halide_start_clock(user_context);
+        }
+    }
+
+    ALWAYS_INLINE ~XRTContext() {
+        (void)halide_xrt_release_context(user_context);  // ignore errors
+    }
+};
+
+// XrtBufferHandle represents a device buffer
+struct XrtBufferHandle {
+    // If nullptr, it means not allocated yet
+    xrtBufferHandle handle;
+    size_t size;
+    bool copy_to_device_pending;
+};
+
+// XrtKernelState represents a loaded kernel on the device
+struct XrtKernelState {
+    xrtKernelHandle handle;
+};
+
+WEAK int create_xrt_context(void *user_context) {
+    unsigned int count = xclProbe();
+    debug(user_context) << "XRT: create_xrt_context: found: " << count << " devices\n";
+
+    if (count == 0) {
+        error(user_context) << "XRT: create_xrt_context: error: no devices were found\n";
+        return halide_error_code_gpu_device_error;
+    }
+
+    for (unsigned i = 0; i < count; i++) {
+        xrtDeviceHandle device = xrtDeviceOpen(i);
+        debug(user_context) << "XRT: create_xrt_context: xrtDeviceOpen: " << device << "\n";
+        if (device != nullptr) {
+            global_device = device;
+            return halide_error_code_success;
+        }
+    }
+
+    return halide_error_code_gpu_device_error;
+}
+
+}  // namespace XRT
+}  // namespace Internal
+}  // namespace Runtime
+}  // namespace Halide
+
+using namespace Halide::Runtime::Internal::XRT;
+
+extern "C" {
+
+WEAK int halide_xrt_device_malloc(void *user_context, halide_buffer_t *buf) {
+    debug(user_context)
+        << "XRT: halide_xrt_device_malloc (user_context: " << user_context
+        << ", buf: " << buf << ")\n";
+
+    if (buf->device) {
+        return halide_error_code_success;
+    }
+
+    XRTContext context(user_context);
+    if (context.error_code)
```

Comment:
Style nit, here and elsewhere: *always* enclose if clauses with {}, even for single statements. Also, do we need to call `error()` here?

Reply:
I have checked other runtime backends and it seems they don't call `error()` here.